component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java

```diff
@@ -0,0 +1,281 @@
+                } catch (IOException | APICallException e) {
+                    log.error("Retry attempt failed.", e);
+                }
+            }
+            response = null;
+        }
+        return response;
+    }
+
+    /**
+     * Will remove Moesif key corresponding to organization ID.
+     * Existing Moesif SDK client associated with the moesif key will be also removed.
+     *
+     * @param orgID
+     */
+    public void removeMoesifKeyFromMap(String orgID) {
+        String moesifKey = orgIDMoesifKeyMap.remove(orgID);
+        moesifKeyClientMap.remove(moesifKey);
+    }
+
+    private void callListResource() throws IOException, APICallException {
+        final URL obj;
+        try {
+            obj = new URL(MoesifMicroserviceConstants.LIST_URL);
+        } catch (MalformedURLException ex) {
+            log.error("Event will be dropped. Getting", ex);
+            return;
+        }
+        String auth = gaAuthUsername + ":" + gaAuthPwd;
+        String encodedAuth = Base64.getEncoder().encodeToString(auth.getBytes(StandardCharsets.UTF_8));
+        String authHeaderValue = "Basic " + encodedAuth;
+        HttpsURLConnection con = null;
+        BufferedReader in = null;
+        try {
+            con = (HttpsURLConnection) obj.openConnection();
+            con.setRequestMethod("GET");
+            con.setRequestProperty("Authorization", authHeaderValue);
+            con.setRequestProperty("Content-Type", MoesifMicroserviceConstants.CONTENT_TYPE);
+            con.setReadTimeout(MoesifMicroserviceConstants.REQUEST_READ_TIMEOUT);
+            int responseCode = con.getResponseCode();
+            if (responseCode == HttpsURLConnection.HTTP_OK) {
+                in = new BufferedReader(new InputStreamReader(con.getInputStream(), "UTF-8"));
+                String inputLine;
+                StringBuffer response = new StringBuffer();
+
+                while ((inputLine = in.readLine()) != null) {
+                    response.append(inputLine);
+                }
+                updateMap(response.toString());
+            } else if (responseCode >= 400 && responseCode < 500) {
+                log.error("Event will be dropped. Getting" + responseCode);
+            } else {
+                throw new APICallException("Getting " + responseCode + " from the microservice and retrying.");
+            }
+        } finally {
+            if (in != null) {
+                in.close();
+            }
+            if (con != null) {
+                con.disconnect();
+            }
+        }
+
+    }
+
+    private String callDetailResource(String orgID) throws IOException, APICallException {
+        StringBuffer response = new StringBuffer();
+        final String url = MoesifMicroserviceConstants.DETAIL_URL_WITH_QUERY + orgID;
+        final URL obj;
+
+        try {
+            obj = new URL(url);
+        } catch (MalformedURLException ex) {
+            log.error("Event will be dropped. Getting", ex);
```

Comment:
This error is not related to an event right. Rather it indicates that it cannot connect to moesif ms. 
In that case lets bring in more meaningful error message indicating that the moesif MS 's url is invalid.

Reply:
Ack.